include/deal.II/base/mpi.h

```diff
@@ -1481,25 +1481,31 @@ namespace Utilities
     {
 #  ifndef DEAL_II_WITH_MPI
       (void)comm;
-      Assert(objects_to_send.size() == 0,
+      Assert(objects_to_send.size() < 2,
              ExcMessage("Cannot send to more than one processor."));
       Assert(objects_to_send.find(0) != objects_to_send.end() ||
                objects_to_send.size() == 0,
              ExcMessage("Can only send to myself or to nobody."));
       return objects_to_send;
 #  else
+      const auto my_proc = this_mpi_process(comm);
 
-      std::vector<unsigned int> send_to(objects_to_send.size());
-      {
-        unsigned int i = 0;
-        for (const auto &m : objects_to_send)
-          send_to[i++] = m.first;
-      }
-      AssertDimension(send_to.size(), objects_to_send.size());
+      std::map<unsigned int, T> received_objects;
+
+      std::vector<unsigned int> send_to;
+      send_to.reserve(objects_to_send.size());
+      for (const auto &m : objects_to_send)
+        if (m.first == my_proc)
+          received_objects[my_proc] = m.second;
+        else
+          send_to.emplace_back(m.first);
+
+      const auto n_point_point_communications =
+        Utilities::MPI::compute_n_point_to_point_communications(comm, send_to);
 
-      const auto receive_from =
-        Utilities::MPI::compute_point_to_point_communication_pattern(comm,
-                                                                     send_to);
+      // Shortcut, when we have nothing to send or nothing to receive
+      if (send_to.size() == 0 && n_point_point_communications == 0)
```

Comment:
```suggestion
      if (send_to.size() == 0 || n_point_point_communications == 0)
```
?

Reply:
Why? At least one of the two blocks below would not be visited with ||, leading to a deadlock.

Reply:
Yes, makes sense. The sentence above confused me. I guess something like
"Shortcut, when we have neither anything to send nor anything to receive" or "Shortcut, when we have nothing to send and nothing to receive"

It's not important to me, though.

Reply:
Thanks!